src/main/scala/zio/keeper/Message.scala

```diff
@@ -1,3 +1,86 @@
 package zio.keeper
 
-final case class Message[S, A](sender: NodeId, state: S, payload: A)
+import java.util.UUID
+
+import zio.keeper.Error.ProtocolDef.SerializationError
+import zio.nio.channels.AsynchronousSocketChannel
+import zio.{ Chunk, IO, ZIO }
+
+import scala.reflect.ClassTag
+
+final case class Message[A](header: Header, payload: A, replyTo: AsynchronousSocketChannel)
+
+case class Header(
+  protocolName: String,
+  protocolVersion: ProtocolDef.Version,
+  sender: NodeId,
+  messageType: Int,
+  payloadSize: Int
+)
+
+object Message {
+  val SIZE_OF_HEADER = 50
+
+  def readHeader(chunk: Chunk[Byte]) = {
```

Comment:
I'd put `readHeader` and `writeHeader` to `Header`'s companion, and rename them to `deserialize` / `serialize` (or whatever notation we're using. I'd do the same for `Message`.

**edit:** `serialize` should live on instance itself not the companion.

Reply:
done. i really don't like adding methods to case classes(to be honest to any classes :))